### DISCUSSION TITLE
Overhaul import strategy

### DISCUSSION THREAD

**Discussion by Haroenv** [algolia/instantsearch#4749] [Ideas]
1. nothing should be added except `version` to `instantsearch` in index.es.ts, those should be imported separately
2. re-export widgets & connectors from index.es.ts, but keep their root folder
3. export things like routers, state mapping, hogan helpers, infinite hits cache form the index.es.ts too
4. change from default exports to named ones for everything (not sure if this is really needed, but it makes `export const connectXXX: xxxConnector` easier than now

Other things that we should change in a major version?

**Comment by francoischalifour** [comment]
Any chance that you give code snippets to what it would look like?

I don't think we should export anything that is coupled with the InstantSearch.js rendering from the root. This goes in line with the long-term "InstantSearch Core" task.

Here's what I've in mind right now:

```js
import instantsearch, { version } from 'instantsearch.js';
import { connectSearchBox } from 'instantsearch.js/connectors';
import { historyRouter } from 'instantsearch.js/routers';
import {
  configure,
  hierarchicalMenu,
  hits,
  pagination,
  panel,
} from 'instantsearch.js/widgets';
```

**Reply by Haroenv** [reply]
This way it makes sense to me as well, however keeping version in the same spot across major versions can be interesting for detection. In our case I don't see that working though, since the version is added to the constructor, and not to the instance to which you'd have access.

Having separate imports for each of the features definitely makes sense, if we find a suitable grouping.

This is what I had in mind:

```js
// mixing default and named exports is very annoying for commonJS, I think I prefer this
import { instantsearch, version } from 'instantsearch.js';
import { connectSearchBox } from 'instantsearch.js/connectors';
import { historyRouter } from 'instantsearch.js/routers';
import {
  simpleStateMapping,
  singleIndexStateMapping,
} from 'instantsearch.js/stateMapping';
import {
  configure,
  hierarchicalMenu,
  hits,
  pagination,
  panel,
} from 'instantsearch.js/widgets';
import {
  snippet,
  reverseSnippet,
  highlight,
  reverseHighlight,
  createInfiniteHitsSessionStorageCache // (or maybe in /widgets?)
} from 'instantsearch.js/helpers'; // maybe "rendering?", not a fan of helpers
import {
  createRoutingMiddleware, // in a major version, we should probably make this the only way of using routing, so tree-shaking can work
  createInsightsMiddleware,
} from 'instantsearch.js/middlewares';
```
